### PR TITLE
Implement an editor to customize the touchscreen controls

### LINKED ISSUE #14456: Android: Allow players to change the button layout
### Problem

As you can see, jump and sneak button blocking my shortcut inventory
Screenshot_20240312-101559

### Solutions

Add custom layout button only for touchscreen device.

### Alternatives

Actually no other alternative

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by grorp**
closes #14456

This PR implements a touchscreen layout editor. Features:

- Move buttons via drag&drop.
- Add/remove buttons. Removed buttons are shown in the grid menu added by #14918 instead.
- Custom layouts are responsive and adapt to changed screen size / DPI / hud_scaling.

https://github.com/user-attachments/assets/202dd8ee-69c5-483b-bad4-9f12ec96c4a7

The UI design is inspired by Minecraft's control editor.

P.S. I have used "ChatGPT" while working on an early version of the touch control editor

## To do

This PR is a Ready for Review.

**Revert commit "Temp: Unbreak mainmenu security" before merging**

Features that would be cool, but are not planned for this initial PR:

- Implement dig/place buttons (#13229). This is already implemented, but the new SVG icons would blow up the diff too much ;)
- Allow having multiple saved layouts, possibly per-game
- Automatically align buttons to each other while dragging, possibly optional
- Make the joystick customizable like a regular button
- Allow resizing buttons

## How to test

Play Minetest on Android. Use the touchscreen layout editor (accessible via the pause menu or the settings menu).

#### APKs for testing (signed with a debug keystore)

[app-armeabi-v7a-release.zip](https://github.com/user-attachments/files/16531491/app-armeabi-v7a-release.zip)
[app-arm64-v8a-release.zip](https://github.com/user-attachments/files/16531479/app-arm64-v8a-release.zip)
[app-x86-release.zip](https://github.com/user-attachments/files/16531495/app-x86-release.zip)
[app-x86_64-release.zip](https://github.com/user-attachments/files/16531499/app-x86_64-release.zip)

The APKs are inside ZIP archives, so you have to extract the ZIP first. You have to delete any existing Minetest installation including data beforehand for installation to work.

### CODE REVIEW

**Comment by RokeJulianLockhart** [comment]
> https://github.com/minetest/minetest/pull/14933#issue-2453114829

@grorp, is there a way to enable a snapping grid, so that buttons can be aligned? I would find such haphazard placement as what that video demonstrates to be unpleasant to use.

**Comment by grorp** [comment]
> is there a way to enable a snapping grid, so that buttons can be aligned?

No, but I might add one in a future PR. See the "Features that would be cool, but are not planned for this initial PR" section.

**Comment by rubenwardy** [comment]
I get the following crash when placing a button from "Add" in the editor. Will try

```
           A  Cmdline: net.minetest.minetest
2024-08-16 20:47:39.943 14490-14490 DEBUG                   pid-14490                            A  pid: 13585, tid: 13681, name: Main  >>> net.minetest.minetest <<<
2024-08-16 20:47:39.943 14490-14490 DEBUG                   pid-14490                            A        #02 pc 0000000000e8b01c  /data/app/~~WFjKVNmSvS5oRum9KTSWCg==/net.minetest.minetest-ph3qyjTNK4rnV5F230Pugg==/lib/arm64/libminetest.so (ButtonMeta::setPos(irr::core::vector2d<int>, irr::core::vector2d<unsigned int>, int)+488) (BuildId: e1349a6a1ba1d4bc3b5a740a43a6cf7345b35bc0)
2024-08-16 20:47:39.943 14490-14490 DEBUG                   pid-14490                            A        #03 pc 0000000000e6a950  /data/app/~~WFjKVNmSvS5oRum9KTSWCg==/net.minetest.minetest-ph3qyjTNK4rnV5F230Pugg==/lib/arm64/libminetest.so (GUITouchscreenLayout::updateDragState(irr::core::vector2d<unsigned int>, irr::core::vector2d<int>)+160) (BuildId: e1349a6a1ba1d4bc3b5a740a43a6cf7345b35bc0)
2024-08-16 20:47:39.943 14490-14490 DEBUG                   pid-14490                            A        #04 pc 0000000000e6b0dc  /data/app/~~WFjKVNmSvS5oRum9KTSWCg==/net.minetest.minetest-ph3qyjTNK4rnV5F230Pugg==/lib/arm64/libminetest.so (GUITouchscreenLayout::OnEvent(irr::SEvent const&)+912) (BuildId: e1349a6a1ba1d4bc3b5a740a43a6cf7345b35bc0)
2024-08-16 20:47:39.943 14490-14490 DEBUG                   pid-14490                            A        #05 pc 0000000000daf484  /data/app/~~WFjKVNmSvS5oRum9KTSWCg==/net.minetest.minetest-ph3qyjTNK4rnV5F230Pugg==/lib/arm64/libminetest.so (irr::gui::IGUIElement::OnEvent(irr::SEvent const&)+68) (BuildId: e1349a6a1ba1d4bc3b5a740a43a6cf7345b35bc0)
2024-08-16 20:47:39.943 14490-14490 DEBUG                   pid-14490                            A        #06 pc 0000000000e5cc4c  /data/app/~~WFjKVNmSvS5oRum9KTSWCg==/net.minetest.minetest-ph3qyjTNK4rnV5F230Pugg==/lib/arm64/libminetest.so (GUIModalMenu::simulateMouseEvent(irr::ETOUCH_INPUT_EVENT, bool)+388) (BuildId: e1349a6a1ba1d4bc3b5a740a43a6cf7345b35bc0)
2024-08-16 20:47:39.943 14490-14490 DEBUG                   pid-14490                            A        #07 pc 0000000000e5d6b8  /data/app/~~WFjKVNmSvS5oRum9KTSWCg==/net.minetest.minetest-ph3qyjTNK4rnV5F230Pugg==/lib/arm64/libminetest.so (GUIModalMenu::preprocessEvent(irr::SEvent const&)+1864) (BuildId: e1349a6a1ba1d4bc3b5a740a43a6cf7345b35bc0)
2024-08-16 20:47:39.943 14490-14490 DEBUG                   pid-14490                            A        #08 pc 0000000000b8b390  /data/app/~~WFjKVNmSvS5oRum9KTSWCg==/net.minetest.minetest-ph3qyjTNK4rnV5F230Pugg==/lib/arm64/libminetest.so (MainMenuManager::preprocessEvent(irr::SEvent const&)+200) (BuildId: e1349a6a1ba1d4bc3b5a740a43a6cf7345b35bc0)
2024-08-16 20:47:39.943 14490-14490 DEBUG                   pid-14490                            A        #09 pc 0000000000c66c38  /data/app/~~WFjKVNmSvS5oRum9KTSWCg==/net.minetest.minetest-ph3qyjTNK4rnV5F230Pugg==/lib/arm64/libminetest.so (MyEventReceiver::OnEvent(irr::SEvent const&)+1056) (BuildId: e1349a6a1ba1d4bc3b5a740a43a6cf7345b35bc0)
2024-08-16 20:47:39.943 14490-14490 DEBUG                   pid-14490                            A        #10 pc 000000000128694c  /data/app/~~WFjKVNmSvS5oRum9KTSWCg==/net.minetest.minetest-ph3qyjTNK4rnV5F230Pugg==/lib/arm64/libminetest.so (irr::CIrrDeviceStub::postEventFromUser(irr::SEvent const&)+72) (BuildId: e1349a6a1ba1d4bc3b5a740a43a6cf7345b35bc0)
2024-08-16 20:47:39.943 14490-14490 DEBUG                   pid-14490                            A        #11 pc 000000000127a2cc  /data/app/~~WFjKVNmSvS5oRum9KTSWCg==/net.minetest.minetest-ph3qyjTNK4rnV5F230Pugg==/lib/arm64/libminetest.so (irr::CIrrDeviceSDL::run()+2544) (BuildId: e1349a6a1ba1d4bc3b5a740a43a6cf7345b35bc0)
2024-08-16 20:47:39.943 14490-14490 DEBUG                   pid-14490                            A        #12 pc 0000000000b3df14  /data/app/~~WFjKVNmSvS5oRum9KTSWCg==/net.minetest.minetest-ph3qyjTNK4rnV5F230Pugg==/lib/arm64/libminetest.so (RenderingEngine::run()+32) (BuildId: e1349a6a1ba1d4bc3b5a740a43a6cf7345b35bc0)
2024-08-16 20:47:39.943 14490-14490 DEBUG                   pid-14490                            A        #13 pc 0000000000c0bfc8  /data/app/~~WFjKVNmSvS5oRum9KTSWCg==/net.minetest.minetest-ph3qyjTNK4rnV5F230Pugg==/lib/arm64/libminetest.so (Game::run()+848) (BuildId: e1349a6a1ba1d4bc3b5a740a43a6cf7345b35bc0)
2024-08-16 20:47:39.943 14490-14490 DEBUG                   pid-14490                            A        #14 pc 0000000000c23b44  /data/app/~~WFjKVNmSvS5oRum9KTSWCg==/net.minetest.minetest-ph3qyjTNK4rnV5F230Pugg==/lib/arm64/libminetest.so (the_game(bool*, InputHandler*, RenderingEngine*, GameStartData const&, std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >&, ChatBackend&, bool*)+128) (BuildId: e1349a6a1ba1d4bc3b5a740a43a6cf7345b35bc0)
2024-08-16 20:47:39.943 14490-14490 DEBUG                   pid-14490                            A        #15 pc 0000000000b8742c  /data/app/~~WFjKVNmSvS5oRum9KTSWCg==/net.minetest.minetest-ph3qyjTNK4rnV5F230Pugg==/lib/arm64/libminetest.so (ClientLauncher::run(GameStartData&, Settings const&)+2576) (BuildId: e1349a6a1ba1d4bc3b5a740a43a6cf7345b35bc0)
2024-08-16 20:47:39.943 14490-14490 DEBUG                   pid-14490                            A        #16 pc 0000000001132b60  /data/app/~~WFjKVNmSvS5oRum9KTSWCg==/net.minetest.minetest-ph3qyjTNK4rnV5F230Pugg==/lib/arm64/libminetest.so (main+2992) (BuildId: e1349a6a1ba1d4bc3b5a740a43a6cf7345b35bc0)
2024-08-16 20:47:39.943 14490-14490 DEBUG                   pid-14490                            A        #17 pc 0000000001199600  /data/app/~~WFjKVNmSvS5oRum9KTSWCg==/net.minetest.minetest-ph3qyjTNK4rnV5F230Pugg==/lib/arm64/libminetest.so (SDL_Main+120) (BuildId: e1349a6a1ba1d4bc3b5a740a43a6cf7345b35bc0)
2024-08-16 20:47:39.944 14490-14490 DEBUG                   pid-14490                            A        #18 pc 00000000018104e0  /data/app/~~WFjKVNmSvS5oRum9KTSWCg==/net.minetest.minetest-ph3qyjTNK4rnV5F230Pugg==/lib/arm64/libminetest.so (Java_org_libsdl_app_SDLActivity_nativeRunMain+712) (BuildId: e1349a6a1ba1d4bc3b5a740a43a6cf7345b35bc0)
2024-08-16 20:47:39.944 14490-14490 DEBUG                   pid-14490                            A        #24 pc 000000000000c500  [anon:dalvik-classes3.dex extracted in memory from /data/app/~~WFjKVNmSvS5oRum9KTSWCg==/net.minetest.minetest-ph3qyjTNK4rnV5F230Pugg==/base.apk!classes3.dex] (org.libsdl.app.SDLMain.run+0)
```

**Comment by grorp** [comment]
> I get the following crash when dragging a button after adding it from "Add" in the editor

Cannot reproduce, but I assume it's this assertion failing:

https://github.com/minetest/minetest/blob/5aeeff0f1415c708d18090d04d0cab53bfbbec40/src/gui/touchscreenlayout.cpp#L137

Replaced it with some printing for debugging, can you tell me what it outputs? And more importantly, does it work now? If it does, I'll just remove the assertion.

**Comment by SmallJoker** [comment]
FWIW: Yesterday I tested this PR on Android 6 and could not reproduce any such crash. Works on my side.

**Comment by rubenwardy** [comment]
```
....
WARNING[Main]: Irrlicht: GL: 8246 8250 Performance:glFlush::Submission has been flushed
WARNING[Main]: Irrlicht: GL: 8246 8250 Performance:glFlush::Submission has been flushed
WARNING[Main]: Irrlicht: GL: 8246 8250 Performance:glFlush::Submission has been flushed
ERROR[Main]: [mismatch]
ERROR[Main]: screensize = (3120,1440)
ERROR[Main]: button size = 193
ERROR[Main]: target pos = (984,732)
ERROR[Main]: actual pos = (983,732)
WARNING[Main]: Irrlicht: GL: 8246 8250 Performance:glFlush::Submission has been flushed
WARNING[Main]: Irrlicht: GL: 8246 8250 Performance:glFlush::Submission has been flushed
WARNING[Main]: Irrlicht: GL: 8246 8250 Performance:glFlush::Submission has been flushed
...
```

**Comment by rubenwardy** [comment]
something for a follow up PR but might be nice to be able to move HUD elements like the minimap and hotbar as part of this: https://github.com/minetest/minetest/issues/15021

although I suppose that is inconsistent with mod HUD

**Comment by Orisphera** [comment]
More items for To Do:

- Make some built-in layouts

File: src/gui/touchscreenlayout.h [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+// Copyright (C) 2024 grorp, Gregor Parzefall <grorp@posteo.de>
+
+#pragma once
+
+#include "client/texturesource.h"
+#include "irr_ptr.h"
+#include "irrlichttypes_bloated.h"
+#include "ITexture.h"
+#include "IGUIStaticText.h"
+#include <unordered_map>
```
Comment (SmallJoker) [inline src/gui/touchscreenlayout.h:12]: `ITexture`, `ISimpleTextureSource` and `IGUIStaticText` are optimally forwards-declared to minimize the size of this header.

**Review by SmallJoker — COMMENTED** [review]
Tested on Android 6.

:ballot_box_with_check: Rearranging buttons
:ballot_box_with_check: Adding buttons
:ballot_box_with_check: Removing buttons
:ballot_box_with_check: Restored layout after app restart

I feel like this menu is hidden too much. It would be great to have it as another entry in the grid menu instead. That's however not something I would require for this PR.

File: builtin/mainmenu/settings/dlg_settings.lua [hunk 2]
```diff
@@ -125,6 +125,22 @@ local function load()
...
 	}
 
+	local touchscreen_layout = {
+		query_text = "Touchscreen layout",
+		requires = {
+			touch_controls = true,
+		},
+		get_formspec = function(self, avail_w)
+			local btn_w = math.min(avail_w, 6)
+			return ("button[0,0;%f,0.8;btn_touch_layout;%s]"):format(btn_w, fgettext("Touchscreen layout")), 0.8
```
[the remark this answers is not in the document]
Comment (grorp) [reply builtin/mainmenu/settings/dlg_settings.lua:135]: It was missing due to a rebase mistake, fixed

**Comment by SmallJoker** [comment]
Cannot test due to #15452.
```
2024-11-18 20:33:51: ERROR[Main]: GUIEngine: execution of menu script failed: Failed to load and run script from /storage/emulated/0/Android/data/net.minetest.minetest/files/Minetest/builtin/init.lua:
2024-11-18 20:33:51: ERROR[Main]: "Mod security: Blocked attempted write to /data/user/0/net.minetest.minetest/cache/cdb"
2024-11-18 20:33:51: ERROR[Main]: 	...netest.minetest/files/Minetest/builtin/mainmenu/init.lua:47: in main chunk
```

I tried to fix it, but at this point it would be easier to have this PR base on an older commit.

**Comment by grorp** [comment]
```diff
diff --cc src/script/scripting_mainmenu.cpp
index d5434c325,d5434c325..3d6c1033d
--- a/src/script/scripting_mainmenu.cpp
+++ b/src/script/scripting_mainmenu.cpp
@@@ -76,26 -76,26 +76,7 @@@ void MainMenuScripting::registerLuaClas
  
  bool MainMenuScripting::mayModifyPath(const std::string &path)
  {
--      if (fs::PathStartsWith(path, fs::TempPath()))
--              return true;
--
--      std::string path_user = fs::RemoveRelativePathComponents(porting::path_user);
--
--      if (fs::PathStartsWith(path, path_user + DIR_DELIM "client"))
--              return true;
--      if (fs::PathStartsWith(path, path_user + DIR_DELIM "games"))
--              return true;
--      if (fs::PathStartsWith(path, path_user + DIR_DELIM "mods"))
--              return true;
--      if (fs::PathStartsWith(path, path_user + DIR_DELIM "textures"))
--              return true;
--      if (fs::PathStartsWith(path, path_user + DIR_DELIM "worlds"))
--              return true;
--
--      if (fs::PathStartsWith(path, fs::RemoveRelativePathComponents(porting::path_cache)))
--              return true;
--
--      return false;
++      return true;
  }
  
  bool MainMenuScripting::checkPathAccess(const std::string &abs_path, bool write_required,
```

**Comment by SmallJoker** [comment]
@grorp I would've done that if I had an Android build system set up at all. Currently I am testing the ARM v7a builds from the buildbot and sign them manually.

**Comment by grorp** [comment]
> Cannot test due to #15452.
> I tried to fix it, but at this point it would be easier to have this PR base on an older commit.

Actually, looks like this PR doesn't even include ea4ae55e24d2c27524490cb8e0d3c34aa46e3da3 yet ...?

File: src/gui/touchscreenlayout.cpp [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+	std::vector<touch_gui_button_id> missing_buttons;
+	for (u8 i = 0; i < touch_gui_button_id_END; i++) {
+		touch_gui_button_id btn = (touch_gui_button_id)i;
+		if (isButtonAllowed(btn) && layout.count(btn) == 0)
+			missing_buttons.push_back(btn);
+	}
+	return missing_buttons;
+}
+
+void ButtonLayout::serializeJson(std::ostream &os) const
```
Comment (SmallJoker) [inline src/gui/touchscreenlayout.cpp:236]: FWIW: We already support groups in `minetest.conf`, which would be a "native" format. This here works too, I suppose - as long nobody ever needs to edit the setting manually (JSON without spaces, indents, or new lines).

I don't think this really needs to be changed - although that's only my opinion.

File: src/gui/touchscreenlayout.cpp [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+bool ButtonLayout::isButtonRequired(touch_gui_button_id id)
+{
+	return id == overflow_id;
+}
+
+s32 ButtonLayout::getButtonSize(v2u32 screensize)
+{
+	return std::min(screensize.Y / 4.5f,
+			RenderingEngine::getDisplayDensity() * 65.0f *
+					g_settings->getFloat("hud_scaling"));
```
Comment (SmallJoker) [inline src/gui/touchscreenlayout.cpp:139]: What's the origin of those magic numbers? Regardless whether random or not: please leave a brief in-line comment.

File: src/gui/touchscreenlayout.cpp [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+}
+
+ButtonLayout ButtonLayout::getDefault()
+{
+	return {{
```
Comment (SmallJoker) [inline src/gui/touchscreenlayout.cpp:144]: This looks like a constant, thus please move it up (if possible). The function is not necessary. Use the constant in `loadFromSettings`.

File: src/gui/touchscreenlayout.cpp [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+			cols++;
+	}
+
+	s32 button_size = ButtonLayout::getButtonSize(screensize);
+	v2s32 spacing(screensize.X / (cols + 1), screensize.Y / (rows + 1));
+	v2s32 pos(spacing);
+
+	for (touch_gui_button_id btn : buttons) {
+		v2u32 orig_size = ButtonLayout::getTexture(btn, tsrc)->getOriginalSize();
+		v2s32 size((f32)orig_size.X / (f32)orig_size.Y * button_size, button_size);
```
Comment (SmallJoker) [inline src/gui/touchscreenlayout.cpp:322]: Trivial: Superfluous use of floating point - unless you'd like to add 0.5f for rounding.

```suggestion
		v2s32 size((button_size * orig_size.X) / orig_size.Y, button_size);
```

+ the same in `ButtonLayout::getRect`.

File: src/script/lua_api/l_mainmenu.cpp [hunk 7]
```diff
@@ -3,6 +3,7 @@
 // Copyright (C) 2013 sapier
 
 #include "lua_api/l_mainmenu.h"
+#include "gui/touchscreeneditor.h"
```
Comment (SmallJoker) [inline src/script/lua_api/l_mainmenu.cpp:6]: Please try to sort new includes a little bit so that it's easier to find duplicates, or relatives.

**Review by SmallJoker — APPROVED** [review]
:ballot_box_with_check: The layout is reset to default if any `Json::RuntimeError` is thrown. The user does not see any error. Luanti continues to work.
:ballot_box_with_check: Move/add/remove buttons from the Settings menu

A few minor comments regarding code quality. Works well as-is.

File: src/gui/touchscreenlayout.cpp (hunk 3 above)
Comment (grorp) [reply src/gui/touchscreenlayout.cpp:236]: JSON has the benefit that you could e.g. share a layout as a file or send it from server to client as well without too much additional work

File: src/gui/touchscreenlayout.cpp (hunk 4 above)
Comment (grorp) [reply src/gui/touchscreenlayout.cpp:139]: These were introduced by b9f618746c0f673ea1f851bb2473fb3ca9410093 it seems.

There's a fixed button size in pixels (adjusted to display density / hud_scaling), and a maximum button size that's proportional to the screen size. But this should be obvious from reading the code, so I'm not sure what I'd write in a comment.

Side note: This code may need adjustment at some point. Currently we have code in `defaultsettings.cpp` that sets different hud_scaling values based on screen size, on Android only. This means that these button size values don't quite work "out of the box" on Linux phones

**Comment by SmallJoker** [comment]
Checked the diff since my last review. Still LGTM. Thanks for the update. :+1:

**Comment by maplemedley** [comment]
Just tested this out on Android, this is such a massive improvement! Tysm for implementing it, this fixes multiple gripes I've had with the mobile experience all at once, and opens up for people to customize their input experience based on their needs! Just being able to remove the zoom button and add the inventory button in the controls is huge for me lol

**Pull request merged**